app.py

```diff
@@ -1,50 +1,150 @@
 import os
+import time
+import json
 
+from redis import Redis
 from flask import (
     Flask, abort, Blueprint, flash, render_template, redirect, request,
-    url_for, flash, jsonify
+    url_for, flash, jsonify, send_from_directory, Markup, session
 )
 from loguru import logger
 from rq import Queue
+from rq.job import Job
 import rq_dashboard
-
 from worker import conn, redis_url
+from flask_socketio import SocketIO
 
 q = Queue(connection=conn)
 
-
-app = Flask(__name__)
+app = Flask(__name__, static_url_path='/static')
 app.config['LOG_LEVEL'] = os.getenv('LOG_LEVEL', 'DEBUG')
 app.config['RQ_DASHBOARD_REDIS_URL'] = redis_url
 app.config.from_object(rq_dashboard.default_settings)
 app.register_blueprint(
     rq_dashboard.blueprint, url_prefix='/rq/'
 )
+app.secret_key = 'test_key'
+
+socket = SocketIO(
+            app,
+            cors_allowed_origins='*',
+            message_queue=os.getenv('REDIS_URL'),
+            manage_session=False
+         ) 
+
+@app.template_filter('empty_session')
+def empty_session(session):
+    '''
+    This function removes all killed/timed-out jobs from the session if the video path for a 
+    given job does not exist. Afterwards, it specifies whether or not the session object is 
+    empty.
+    Parameters:
+    - session: Flask session object
+    '''
+   
+    if 'jobs' not in session:
+        logger.info('Empty session')
+        return True
+
+    def clear_session():
+        for idx, job_dict in enumerate(session['jobs']):
+            job = q.fetch_job(job_dict['id'])
+            video_source = 'static/videos/{}'.format(job_dict['tracks_filename'])
+            
+            dead_job_or_timeout = job is None or job.get_status() == 'failed'
+            video_does_not_exist = not os.path.exists(video_source)
+            
+            if dead_job_or_timeout and video_does_not_exist:
+                session['jobs'].pop(idx)
+                # Have to set modified to True after every in-place operation
+                session.modified = True 
+                logger.info('Popped job from session: id={}'.format(job_dict['id']))
+
+    clear_session()
+    
+    if len(session['jobs']) == 0:
+        logger.info('Empty session')
+        return True
+    
+    return False
 
-jobs = []
+@app.template_filter('refresh_job')
+def refresh_job(job_dict):
+    '''
+    This function updates the meta dictionary of a given job and returns the
+    job's filename
+    Parameters:
+    - job_dict: Dictionary containing a job id and job filename
+    '''
+    logger.info('Refresh job, id={}'.format(job_dict['id']))
+    job = q.fetch_job(job_dict['id'])
+    filename = job_dict['filename']
+    try:
+        job.refresh()
+        logger.info('Job refreshed successfully')
+    except Exception as e: 
```

Comment:
abrichr 9 days ago Member
Please specify which exception you are catching and print it to stdout.